File: frontend/src/types/analytics.ts
export interface AnalyticsParams {
  date_from?: string
  date_to?: string
  asset_id?: string
  location_id?: string
  category_id?: string
  granularity?: 'day' | 'week'
}

export interface CountRow {
  asset_id: string | null
  user_id: string | null
  category_id: string | null
  count: number
}

export interface WorkOrderAnalytics {
  date_from: string
  date_to: string
  total: number
  by_status: Record<string, number>
  by_priority: Record<string, number>
  completed: number
  completion_rate: number
  overdue: number
  avg_cycle_time_hours: number | null
  avg_response_time_hours: number | null
  by_asset: CountRow[]
  by_user: CountRow[]
  by_category: CountRow[]
}

export interface RequestAnalytics {
  date_from: string
  date_to: string
  total: number
  by_status: Record<string, number>
  by_priority: Record<string, number>
  received: number
  resolved: number
  converted: number
  avg_resolution_cycle_hours: number | null
}

export interface PartCostRow {
  part_id: string
  custom_id: string
  name: string
  qty: string
  cost: string
}
export interface AssetCostRow {
  asset_id: string | null
  cost: string
}
export interface VendorSpendRow {
  vendor_id: string
  spend: string
}
export interface MaintenanceCostByAssetRow {
  asset_id: string | null
  parts_cost: string
  labor_cost: string
  additional_cost: string
  total: string
}
export interface CostAnalytics {
  date_from: string
  date_to: string
  parts_consumption_cost: string
  consumption_by_part: PartCostRow[]
  consumption_by_asset: AssetCostRow[]
  po_spend_approved: string
  po_spend_by_vendor: VendorSpendRow[]
  labor_cost: string
  additional_cost: string
  total_maintenance_cost: string
  maintenance_cost_by_asset: MaintenanceCostByAssetRow[]
}

export interface AssetReliabilityRow {
  asset_id: string
  custom_id: string
  name: string
  availability_pct: number
  downtime_count: number
  total_downtime_hours: number
  mttr_hours: number | null
  mtbf_hours: number | null
  total_maintenance_cost: string
  acquisition_cost: string | null
  cost_to_value_ratio: number | null
}
export interface AssetReliabilityAnalytics {
  date_from: string
  date_to: string
  window_hours: number
  assets: AssetReliabilityRow[]
  fleet_availability_pct: number | null
  fleet_total_downtime_hours: number
  fleet_mttr_hours: number | null
  fleet_mtbf_hours: number | null
  fleet_total_maintenance_cost: string
}

export interface CategoryValueRow {
  category_id: string | null
  name: string | null
  value: string
}
export interface LowStockRow {
  part_id: string
  custom_id: string
  name: string
  quantity: string
  min_quantity: string
  shortfall: string
}
export interface TopConsumedRow {
  part_id: string
  custom_id: string
  name: string
  qty: string
}
export interface ABCRow {
  part_id: string
  custom_id: string
  name: string
  consumption_value: string
  cumulative_pct: number
  abc_class: string
}
export interface InventoryAnalytics {
  total_inventory_value: string
  inventory_value_by_category: CategoryValueRow[]
  low_stock_count: number
  low_stock_items: LowStockRow[]
  top_consumed_parts: TopConsumedRow[]
  abc_classification: ABCRow[]
  abc_summary: Record<string, number>
}

export interface PersonnelRow {
  user_id: string
  name: string | null
  created_count: number
  completed_count: number
  assigned_count: number
  labor_hours: number
  labor_cost: string
}
export interface PersonnelAnalytics {
  date_from: string
  date_to: string
  users: PersonnelRow[]
}

export interface TrendBucket {
  bucket_start: string
  work_orders_created: number
  work_orders_completed: number
  requests_received: number
  requests_resolved: number
}
export interface TrendAnalytics {
  date_from: string
  date_to: string
  granularity: string
  buckets: TrendBucket[]
}

File: backend/app/schemas/analytics.py
"""分析仪表盘响应 schema（Phase 4，只读）。"""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class CountRow(BaseModel):
    asset_id: str | None = None
    user_id: str | None = None
    category_id: str | None = None
    count: int


class WorkOrderAnalytics(BaseModel):
    date_from: date
    date_to: date
    total: int
    by_status: dict[str, int]
    by_priority: dict[str, int]
    completed: int
    completion_rate: float
    overdue: int
    avg_cycle_time_hours: float | None
    avg_response_time_hours: float | None
    by_asset: list[CountRow]
    by_user: list[CountRow]
    by_category: list[CountRow]


class RequestAnalytics(BaseModel):
    date_from: date
    date_to: date
    total: int
    by_status: dict[str, int]
    by_priority: dict[str, int]
    received: int
    resolved: int
    converted: int
    avg_resolution_cycle_hours: float | None


class PartCostRow(BaseModel):
    part_id: str
    custom_id: str
    name: str
    qty: Decimal
    cost: Decimal


class AssetCostRow(BaseModel):
    asset_id: str | None
    cost: Decimal


class VendorSpendRow(BaseModel):
    vendor_id: str
    spend: Decimal


class MaintenanceCostByAssetRow(BaseModel):
    asset_id: str | None
    parts_cost: Decimal
    labor_cost: Decimal
    additional_cost: Decimal
    total: Decimal


class CostAnalytics(BaseModel):
    date_from: date
    date_to: date
    parts_consumption_cost: Decimal
    consumption_by_part: list[PartCostRow]
    consumption_by_asset: list[AssetCostRow]
    po_spend_approved: Decimal
    po_spend_by_vendor: list[VendorSpendRow]
    labor_cost: Decimal
    additional_cost: Decimal
    total_maintenance_cost: Decimal
    maintenance_cost_by_asset: list[MaintenanceCostByAssetRow]


class AssetReliabilityRow(BaseModel):
    asset_id: str
    custom_id: str
    name: str
    availability_pct: float
    downtime_count: int
    total_downtime_hours: float
    mttr_hours: float | None
    mtbf_hours: float | None
    total_maintenance_cost: Decimal
    acquisition_cost: Decimal | None
    cost_to_value_ratio: float | None


class AssetReliabilityAnalytics(BaseModel):
    date_from: date
    date_to: date
    window_hours: float
    assets: list[AssetReliabilityRow]
    fleet_availability_pct: float | None
    fleet_total_downtime_hours: float
    fleet_mttr_hours: float | None
    fleet_mtbf_hours: float | None
    fleet_total_maintenance_cost: Decimal


class CategoryValueRow(BaseModel):
    category_id: str | None
    name: str | None
    value: Decimal


class LowStockRow(BaseModel):
    part_id: str
    custom_id: str
    name: str
    quantity: Decimal
    min_quantity: Decimal
    shortfall: Decimal


class TopConsumedRow(BaseModel):
    part_id: str
    custom_id: str
    name: str
    qty: Decimal


class ABCRow(BaseModel):
    part_id: str
    custom_id: str
    name: str
    consumption_value: Decimal
    cumulative_pct: float
    abc_class: str


class PersonnelRow(BaseModel):
    user_id: str
    name: str | None
    created_count: int
    completed_count: int
    assigned_count: int
    labor_hours: float
    labor_cost: Decimal


class PersonnelAnalytics(BaseModel):
    date_from: date
    date_to: date
    users: list[PersonnelRow]


class TrendBucket(BaseModel):
    bucket_start: date
    work_orders_created: int
    work_orders_completed: int
    requests_received: int
    requests_resolved: int


class TrendAnalytics(BaseModel):
    date_from: date
    date_to: date
    granularity: str
    buckets: list[TrendBucket]


class WoCategoryConsumptionRow(BaseModel):
    category_id: str | None
    name: str | None
    cost: Decimal
    qty: Decimal


class MonthlyConsumptionRow(BaseModel):
    month: str
    cost: Decimal


class InventoryAnalytics(BaseModel):
    total_inventory_value: Decimal
    inventory_value_by_category: list[CategoryValueRow]
    low_stock_count: int
    low_stock_items: list[LowStockRow]
    top_consumed_parts: list[TopConsumedRow]
    abc_classification: list[ABCRow]
    abc_summary: dict[str, int]
    consumption_by_wo_category: list[WoCategoryConsumptionRow] = []
    consumption_monthly_trend: list[MonthlyConsumptionRow] = []
